All analyses table

dmfalke opened the pull request:
### Summary
This PR introduces the **All Analyses** table, including the ability to "pin" analyses to the top of the table, and to remove unpinned analyses.

### Detailed changes
* Methods for getting/setting Analyses preferences
* Method to delete multiple analyses
* `useAnalysisList` React hook
* `usePinnedAnalyses` React hook
* `AllAnalyses` React Component
* Modified root route to render `AllAnalyses`

### Related issues
- closes #292

bobular commented:
Hi @dmfalke 
I appreciate that the entry point to each study might be from the home page cards or other site navigation (menus), but if we merge this branch, we (developers) will have no way to start a new analysis for a study that's not already listed in the My Analyses table.

Can we bring back the list of studies?  Even if just for devs.

dmfalke commented:
@bobular, yes, that's a good idea. I will add that.

bobular commented on src/lib/workspace/AllAnalyses.tsx:
```diff
@@ -0,0 +1,338 @@
+} from '@material-ui/core';
+import { Loading, Mesa } from '@veupathdb/wdk-client/lib/Components';
+import { ContentError } from '@veupathdb/wdk-client/lib/Components/PageStatus/ContentError';
+import { useSessionBackedState } from '@veupathdb/wdk-client/lib/Hooks/SessionBackedState';
```
What's the difference between this and the localforage thing we're using in Mocks.ts ?

dmfalke replied:
`localforage` uses a more persistent storage by default (typically [IndexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API)). `useSessionBackedState` uses a storage that gets cleared when the tab/window is closed ([sessionState](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage)).

bobular commented on src/lib/workspace/WorkspaceRouter.tsx:
```diff
@@ -38,6 +47,11 @@ export function WorkspaceRouter({
+      <Route
+        path={`${path}/studies`}
+        exact
+        render={() => <StudyList subsettingServiceUrl={subsettingServiceUrl} />}
```
When called from this route, StudyList is rendering links with URLs like `/eda/studies/DS_841ea82`, but they just need to be `/eda/DS_841ea82`

dmfalke replied:
Ah, right. I will fix.

bobular commented on src/lib/workspace/AnalysisList.tsx:
```diff
@@ -14,16 +20,11 @@ export function AnalysisList(props: Props) {
-  React.useEffect(() => {
-    updateAnalysisList();
-  }, [updateAnalysisList]);
+  const { analyses: analysisList, deleteAnalyses } = useAnalysisList(
+    analysisStore
+  );
```
Need to reinstate the filter on studyId 
```suggestion
  const { analyses, deleteAnalyses } = useAnalysisList(
    analysisStore
  );
  const analysisList = analyses ? analyses.filter((analysis) => analysis.studyId === studyId) : undefined;
```

bobular commented on src/lib/workspace/AllAnalyses.tsx:
```diff
@@ -0,0 +1,338 @@
+                    return (
+                      datasets?.records.find(
+                        (d) => d.id[0].value === analysis.studyId
+                      )?.displayName ?? 'Unknown study'
```
There's a second or so's delay when I reload this page, and "Unknown study" is a bit disconcerting.

Could this work?
```suggestion
                      )?.displayName ?? 'Loading...'
```

bobular replied:
Actually this is the wrong place to do this (see below)

bobular commented on src/lib/workspace/AllAnalyses.tsx:
```diff
@@ -0,0 +1,338 @@
+            const dataset = datasets?.records.find(
+              (d) => d.id[0].value === data.row.studyId
+            );
+            if (dataset == null) return 'Unknown study';
```
```suggestion
            if (dataset == null) return 'Loading...';
```

dmfalke replied:
I'd prefer to defer rendering of the table until the datasets are loaded. I will update in a bit.

bobular commented on src/lib/workspace/AllAnalyses.tsx:
```diff
@@ -0,0 +1,338 @@
+            if (dataset == null) return 'Unknown study';
+            return (
+              <Link to={`${url}/${dataset.id[0].value}`}>
+                {dataset.displayName}
```
SCORE S. mansoni italics tags are currently showing.  The approach in StudyList can be used here to clean the HTML.

```
import { safeHtml } from '@veupathdb/wdk-client/lib/Utils/ComponentUtils';
```

```suggestion
                {safeHtml(dataset.displayName)}
```

dmfalke replied:
Thanks, this is good! I will update soon.

bobular reviewed (approved):
Did some thorough testing and found the bug with the study-page's AnalysisList.  Fixes and suggestions provided above/below.

dmfalke commented:
I've made some updates to this branch:
- Updated datetime formatting to be consistent with other ClinEpiDB pages
- Rearranged layout of table actions
- Added search box for table
- Added confirmation modal for delete actions
- Only render after datasets have been loaded to prevent **Unknown study** in all rows

jtlong3rd commented:
> Just pushed a fix for some type bugs. This is a result of `Mesa` being up-typed 🙁,

Story of my (VEuPathDB) life!

dmfalke commented:
> > Just pushed a fix for some type bugs. This is a result of `Mesa` being up-typed slightly_frowning_face,
> 
> Story of my (VEuPathDB) life!

word

bobular commented:
> Only render after datasets have been loaded to prevent **Unknown study** in all rows

This is great, but it now needs a progress spinner IMO!

bobular commented:
> > Only render after datasets have been loaded to prevent **Unknown study** in all rows
> 
> This is great, but it now needs a progress spinner IMO!

Hmm, I see there's this
```
        {(loading || datasets == null) && <Loading />}
```
It's nested inside the `{analyses && datasets && ( ...` so needs to go outside of this.

bobular commented on src/lib/workspace/AllAnalyses.tsx:
```diff
@@ -0,0 +1,423 @@
+  const theme = createMuiTheme(workspaceTheme);
+  return (
+    <ThemeProvider theme={theme}>
+      <div className={classes.root}>
+        <h1>My Analyses</h1>
+        {error && <ContentError>{error}</ContentError>}
+        {analyses && datasets && (
+          <Mesa.Mesa state={tableState}>
+            <div
+              style={{
+                display: 'flex',
+                alignItems: 'center',
+                gap: '1ex',
+              }}
+            >
+              <TextField
+                variant="outlined"
+                size="small"
+                label="Search analyses"
+                inputProps={{ size: 50 }}
+                value={searchText}
+                onChange={(e) => setSearchText(e.target.value)}
+              />
+              <span>
+                Showing {filteredAnalysesAndDatasets?.length} of{' '}
+                {analyses.length} analyses
+              </span>
+            </div>
+            {(loading || datasets == null) && <Loading />}
+          </Mesa.Mesa>
+        )}
+      </div>
```
```suggestion
      <div className={classes.root}>
        <h1>My Analyses</h1>
        {error && <ContentError>{error}</ContentError>}
        {(loading || datasets == null) && <Loading />}
        {analyses && datasets && (
          <Mesa.Mesa state={tableState}>
            <div
              style={{
                display: 'flex',
                alignItems: 'center',
                gap: '1ex',
              }}
            >
              <TextField
                variant="outlined"
                size="small"
                label="Search analyses"
                inputProps={{ size: 50 }}
                value={searchText}
                onChange={(e) => setSearchText(e.target.value)}
              />
              <span>
                Showing {filteredAnalysesAndDatasets?.length} of{' '}
                {analyses.length} analyses
              </span>
            </div>
          </Mesa.Mesa>
        )}
      </div>
```

bobular replied:
Sorry, that's a much bigger-looking diff than it really is!

I moved the `<Loading />` line up to the top :-)

bobular reviewed (approved):
Looks great - just one tiny fix suggested.

dmfalke commented:
@bobular I made a change close to what you suggested. I'm going to go ahead and merge this now :slightly_smiling_face: .
